config/src/main/java/org/springframework/security/config/annotation/method/configuration/ReactiveMethodSecurityConfiguration.java

```diff
@@ -16,87 +16,89 @@
 
 package org.springframework.security.config.annotation.method.configuration;
 
-import java.util.Arrays;
-
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.config.BeanDefinition;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
-import org.springframework.context.annotation.ImportAware;
 import org.springframework.context.annotation.Role;
-import org.springframework.core.type.AnnotationMetadata;
 import org.springframework.security.access.expression.method.DefaultMethodSecurityExpressionHandler;
-import org.springframework.security.access.expression.method.ExpressionBasedAnnotationAttributeFactory;
-import org.springframework.security.access.expression.method.ExpressionBasedPostInvocationAdvice;
-import org.springframework.security.access.expression.method.ExpressionBasedPreInvocationAdvice;
 import org.springframework.security.access.expression.method.MethodSecurityExpressionHandler;
-import org.springframework.security.access.intercept.aopalliance.MethodSecurityMetadataSourceAdvisor;
-import org.springframework.security.access.method.AbstractMethodSecurityMetadataSource;
-import org.springframework.security.access.method.DelegatingMethodSecurityMetadataSource;
-import org.springframework.security.access.prepost.PrePostAdviceReactiveMethodInterceptor;
-import org.springframework.security.access.prepost.PrePostAnnotationSecurityMetadataSource;
+import org.springframework.security.authorization.method.PostAuthorizeReactiveAuthorizationManager;
+import org.springframework.security.authorization.method.PostFilterAuthorizationAfterReactiveMethodInterceptor;
+import org.springframework.security.authorization.method.PreAuthorizeReactiveAuthorizationManager;
+import org.springframework.security.authorization.method.PreFilterAuthorizationBeforeReactiveMethodInterceptor;
+import org.springframework.security.authorization.method.ReactiveAuthorizationAfterMethodInterceptor;
+import org.springframework.security.authorization.method.ReactiveAuthorizationBeforeMethodInterceptor;
+import org.springframework.security.authorization.method.ReactiveAuthorizationManagerAfterMethodInterceptor;
+import org.springframework.security.authorization.method.ReactiveAuthorizationManagerBeforeMethodInterceptor;
 import org.springframework.security.config.core.GrantedAuthorityDefaults;
 
 /**
  * @author Rob Winch
  * @author Tadaya Tsuyukubo
+ * @author Evgeniy Cheban
  * @since 5.0
  */
 @Configuration(proxyBeanMethods = false)
-class ReactiveMethodSecurityConfiguration implements ImportAware {
-
-	private int advisorOrder;
+class ReactiveMethodSecurityConfiguration {
 
-	private GrantedAuthorityDefaults grantedAuthorityDefaults;
+	@Bean
+	@Role(BeanDefinition.ROLE_INFRASTRUCTURE)
+	ReactiveAuthorizationBeforeMethodInterceptor reactiveAuthorizationBeforeMethodInterceptor() {
```

Comment:
While this solution does have fewer configuration points, it's still not as simple as the imperative-based interceptors.

Have you already tried having each method interceptor stand on its own without intervening support? This may result in a small amount of copy-paste between the interceptors, but it comes with the nice benefits that the imperative interceptors have.

Reply:
Sorry, I don't quite understand your idea. Can you describe it in more detail?

Reply:
I mean introducing `AuthorizationManagerBeforeReactiveMethodInterceptor` that does all the pre-advice work found in `PrePostAdviceReactiveMethodInterceptor` and `AuthorizationManagerAfterReactiveMethodInterceptor` that does all the post-advice work found there. 

The pre-advice code in `PrePostAdviceReactiveMethodInterceptor` would be duplicated in `AuthorizationManagerBeforeReactiveMethodInterceptor` and `PreFilterAuthorizationReactiveMethodInterceptor`, for example.

In this way, an application only needs an `AuthorizationManagerBeforeReactiveMethodInterceptor` to configure for `@PreAuthorize`. Not an `AuthorizationManagerBeforeReactiveMethodInterceptor`, an `AuthorizationBeforeReactiveMethodInterceptor`, and an `AuthorizationAfterReactiveMethodInterceptor`.

I did a sketch of this in a separate branch to illustrate: https://github.com/jzheaux/spring-security/blob/gh-9401/core/src/main/java/org/springframework/security/authorization/method/AuthorizationManagerBeforeReactiveMethodInterceptor.java

Reply:
The problem seems to be that we are calling `CoroutinesUtils.invokeSuspendingFunction` when `isSuspendingFunction` is `true` so the rest of the interceptors will not be called. In my current version the `CoroutinesUtils.invokeSuspendingFunction` will be called only after all interceptors are executed.

I added the `AuthorizationReactiveMethodInterceptorBeanFactoryPostProcessor` that dynamically registers the `AuthorizationBeforeReactiveMethodInterceptor` and the `AuthorizationAfterReactiveMethodInterceptor` beans. The `AuthorizationManagerBeforeReactiveMethodInterceptor` implements `BeanDefinitionRegistryPostProcessor`, so it will add the required bean definitions to the `BeanDefinitionRegistry`.

@jzheaux I've updated the PR, please take a look.